### PR TITLE
[Interpreter] Change-based-sim and misc feature round-up

### PR DESCRIPTION
**Pull request by EclecticGriffin**
Sorry for the eclectic PR, kinda kept jumping from one thing to another on this one. This takes care of a few things that have been on the back of my todo list for a while:
- removes the outmoded `-p` flag from the debugger
- Remove the global config struct and weave the configuration options through the appropriate calls
- Update the root logger to feature an explicit initiation call and replace use of `lazy_static` with `once_cell` to make the deferred initialization easier
- Add an overflow warning to the `StdFpAdd` and `StdFpSub` primitives. Not sure why this was missing in the first place.
- Added a `ConfigBuilder` struct to make putting together the configuration a little nicer. It's not really necessary for anything but it does make me happy, so there's that at least.

This also adds the changed based simulation under a rust feature called `change-based-sim`. You can build the interpreter with it via:
```
cargo build --features change-based-sim
```

or (from the root dir)
```
cargo build --features interp/change-based-sim
```

Numbers to follow soon

### CODE REVIEW

File: interp/src/macros.rs [hunk 1]
```diff
@@ -17,6 +17,7 @@
 /// `StdAdd::new(bindings: ir::Params)` and `StdAdd::from_constants(ports)`
 ///
 /// TODO(rachit): $out_width is never used.
+/// (TODO: Griffin) fix the copy-paste
```
Comment (rachitnigam) [inline interp/src/macros.rs:20]: Does this need to be addressed before merge?

Comment (EclecticGriffin) [reply interp/src/macros.rs:20]: No, it's just gross and should probably be better in the future.

**Comment by EclecticGriffin** [comment]
Crunched the benchmark numbers and it seems like it is marginally faster for all the benchmarks. This doesn't mean a whole lot for the main benchmark suite as they all run rather quickly, so I didn't bother to check for statistical significance.

But I'm sure everyone is interested in LeNet which gave us the following:
```
simulation,stage,mean,median,stddev
LeNet,interpreter,2430.823,2408.577,91.761          (The original)
LeNet,interpreter,2199.59,2195.994,24.319            (Change based)
```

Time is in seconds.

Though I technically didn't run the current interpreter without change-based, so this comparison is against the original benchmark numbers which means that it could possibly be something else. I intend to get numbers for the current version of the interpreter without change-based sim enabled, but haven't done so yet and think it's probably not worth holding this PR back over

**Comment by rachitnigam** [comment]
Big change since the original benchmarks is the dataflow order pass which could have a similar effect. Real test is in randomly sorting the assignments and seeing if change based helps.

**Comment by EclecticGriffin** [comment]
Running with the old benchmarks sans dataflow ordering and the ordering itself won't affect either change-based or non-change-based since the assignments are still interpreted in parallel

**Comment by EclecticGriffin** [comment]
Gonna merge this now if there are no objections

**Comment by rachitnigam** [comment]
> Running with the old benchmarks sans dataflow ordering and the ordering itself won't affect either change-based or non-change-based since the assignments are still interpreted in parallel

The point of this would be to quantify the utility of change-based simulation. We already expect it to be "better" but better by how much?

Merging sounds good. We should run LeNet with default settings and actually get a sense of how much faster it is.

**Comment by EclecticGriffin** [comment]
For reference lenet running with the normal interpretation on the current build of the interpreter produces:

```
LeNet,interpreter,2364.823,2365.88,15.837
```

**Comment by sampsyo** [comment]
Neato!!!!!! Awesome job making this work!

To summarize the empirical results, the data tells us that the "exhaustive" mode on LeNet took 39m20s ± 16s on average while the optimized "change-based" mode took 36m40s ± 24 seconds (using an appropriate number of significant figures). So that's a savings of about 6.8% of the execution time (while the standard deviation is around 0.7% of the execution time).

That's definitely a win, but in a way I'm a tad surprised it's not a larger win. I think it would be super useful to dig a little bit into the underlying reasons. Two ideas for metrics come to mind:

* I can think of one big time-independent metric we could collect (which could be revealing on even much smaller, non-LeNet programs): how many assignments actually ran? Understanding how many assignment evaluations were skipped could be super interesting.
* What is the cost of checking for changes? The change-based mode could have an option to check for changes, but actually run everything anyway! The excess cost would then reflect the overhead of change tracking.

Beyond these specific metrics, it could be very instructive to just get a profiling breakdown of one LeNet run… seeing where the time goes in the two configurations could be super instructive. For example, it could be cool to see if the change checking shows up in a meaningful way on the profile.

**Pull request merged**